### PR TITLE
Basic caching

### PR DESCRIPTION
**Pull request by npmaile**
this should implement a basic caching scheme to make card operations quite a bit faster.

### CODE REVIEW

File: gui/api.go [hunk 1]
```diff
@@ -314,19 +347,32 @@ func (apiSession apiSession) listPhonons(w http.ResponseWriter, r *http.Request)
...
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	for _, p := range phonons {
-		p.PubKey, err = sess.GetPhononPubKey(p.KeyIndex)
+
+	phonons := []*model.Phonon{}
+	if cache[sess.GetName()].cachePopulated {
+		phonons = []*model.Phonon{}
```
Comment (maevyn11) [inline gui/api.go:353]: Where does cachePopulated get set? I don't see it in createPhonon, initDeposit, or finalizeDeposit.

File: gui/api.go [hunk 2]
```diff
@@ -314,19 +347,32 @@ func (apiSession apiSession) listPhonons(w http.ResponseWriter, r *http.Request)
...
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		}
+		for _, phonon := range phonons {
+			cache[sess.GetName()].phonons[phonon.KeyIndex] = phonon
+		}
+		cache[sess.GetName()].cachePopulated = true
+	}
```
Comment (maevyn11) [inline gui/api.go:375]: OK, I see. So this just caches lists after the first one.

File: gui/api.go [hunk 3]
```diff
@@ -375,7 +421,7 @@ func (apiSession apiSession) setDescriptor(w http.ResponseWriter, r *http.Reques
 		CurrencyType: model.CurrencyType(inputs.CurrencyType),
 	}
 	p.KeyIndex = uint16(index)
-
+	cache[sess.GetName()].phonons[p.KeyIndex] = p
 	err = sess.SetDescriptor(p)
```
Comment (maevyn11) [inline gui/api.go:425]: seems like this should maybe be after the error is checked on `err = sess.SetDescriptor`. If that fails then the cache and card will be out of sync for this keyIndex.

File: gui/api.go [hunk 4]
```diff
@@ -400,7 +446,10 @@ func (apiSession apiSession) send(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "Unable to convert index to int:"+err.Error(), http.StatusBadRequest)
 		return
 	}
+
 	err = sess.SendPhonons([]uint16{uint16(index)})
+	delete(cache[sess.GetName()].phonons, uint16(index))
+
```
Comment (maevyn11) [inline gui/api.go:452]: again this should probably be after the error check. If the key fails to send the cache will be out of sync.

**Review by maevyn11 — CHANGES_REQUESTED** [review]
Looks good for the most part. Just a couple small comments that I think will make it more robust.

File: gui/api.go (hunk 2 above)
Comment (npmaile) [reply gui/api.go:375]: yeah, the idea is that the cache isn't fully populated until a full list has been done at least one time. After that anything that would modify the cache is modified in the cache along with the on card change happening

File: gui/api.go (hunk 3 above)
Comment (npmaile) [reply gui/api.go:424]: fixed

File: gui/api.go (hunk 4 above)
Comment (npmaile) [reply gui/api.go:451]: fixed

**Review by maevyn11 — APPROVED** [review]
LOOKS GOOD

**Pull request merged**